src/libostree/ostree-repo-pull.c

```diff
@@ -2006,6 +1972,105 @@ _ostree_preload_metadata_file (OstreeRepo    *self,
 }
 
 static gboolean
+fetch_mirrorlist (OstreeFetcher  *fetcher,
+                  const char     *mirrorlist_url,
+                  GSList        **out_mirrorlist,
+                  GCancellable   *cancellable,
+                  GError        **error)
+{
+  gboolean ret = FALSE;
+  char **lines = NULL;
+  g_autofree char *contents = NULL;
+  GSList *ret_mirrorlist = NULL;
+  SoupURI *mirrorlist = NULL;
+
+  mirrorlist = soup_uri_new (mirrorlist_url);
+  if (mirrorlist == NULL)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "Failed to parse mirrorlist URL '%s'", mirrorlist_url);
+      goto out;
+    }
+
+  if (!fetch_uri_contents_utf8_sync (fetcher, mirrorlist, &contents,
+                                     cancellable, error))
+    {
+      g_prefix_error (error, "While fetching mirrorlist '%s': ",
+                      mirrorlist_url);
+      goto out;
+    }
+
+  /* go through each mirror in mirrorlist and do a quick sanity check that it
+   * works so that we don't waste the fetcher's time when it goes through them.
```

Comment:
Yeah, that's a good point. OTOH, if the first few mirrors are bad, then it's faster overall to do the check here. I fine-tuned this in the last fixup commit so that we only check up to the first happy mirror.